### DISCUSSION TITLE
Refresh Problems on Mac

### DISCUSSION THREAD

**Discussion by MichaelWDenney** [kirill-grouchnikov/radiance#492] [Q&A]
I realize this is a very general question, but can anyone suggest any cause for refresh problems on Macs?  On Macs there are a lot of controls, both in the ribbons and in our own interfaces wrapped by the ribbon frame, that fail to refresh.  Dragging the cursor over these controls, forces a repaint.  Our GUI refreshes fine on Windows (which is were most of the GUI development occurs).

**Comment by kirill-grouchnikov** [comment]
Any specific ones, or across the board?

**Comment by MichaelWDenney** [comment]
It is more or less across the board.  For example, if I shrink the GUI by dragging the one side in and then drag it back out, must of the GUI that was not visible when the GUI was shrunk will be blank with a bluish background.  After a moment, the ribbon will redraw, but the tittle bar and all our components  (in the equivalent location of the text pane and control panel in the ribbon demo) remain blanked out.

I've seen popups that take a slight delay to draw and at least one that never fully refreshed properly.

I was also told by another tester than when they moved the GUI to a different display the entire interface went blank with the bluish background.  In that case, even the ribbon doesn't appear to have eventually redrawn properly.
![image](https://github.com/user-attachments/assets/c677f70e-402e-4a0c-9dbd-9f53c4bfdfca)

Again, I have never seen these problems on Windows.  You work on a Mac, correct?  I assume you haven't seen this behaviour.  I wonder if I've made a fundamental error somewhere.

**Reply by kirill-grouchnikov** [reply]
This is probably beyond Radiance and into the specific rendering pipeline on that version of macOS under that version of JDK. 

Try one or both of these maybe? Just shooting in the dark here really.

```
-Dsun.java2d.d3d=false  -Dsun.java2d.noddraw=true

System.setProperty("sun.java2d.d3d", "false");
System.setProperty("sun.java2d.noddraw", "true");
```

D3D should only be relevant for Windows, I think.

**Reply by kirill-grouchnikov** [reply]
There's a bunch more at https://docs.oracle.com/javase/8/docs/technotes/guides/2d/flags.html

**Reply by MichaelWDenney** [reply]
Still having refresh/redraw problems on Macs.  We are receiving the following exception:

```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException: Cannot invoke "javax.swing.border.Border.paintBorder(java.awt.Component, java.awt.Graphics, int, int, int, int)" because the return value of "com.apple.laf.AquaMenuPainter$RecyclableBorder.get()" is null
	at java.desktop/com.apple.laf.AquaMenuPainter.paintMenuBarBackground(AquaMenuPainter.java:149)
```

Any ideas?

**Reply by kirill-grouchnikov** [reply]
No, since it's in Aqua menu classes. Are you following the steps in the "How can I use Radiance on Mac and still have the standard Mac application menu bar?" section of https://github.com/kirill-grouchnikov/radiance/blob/sunshine/docs/faq.md? It's been a very long time since that one was written, and I've never tested that in Radiance test apps.

**Reply by MichaelWDenney** [reply]
No.  That was up for consideration, but at this point we are just using the ribbon interface on all platforms.

**Reply by kirill-grouchnikov** [reply]
Then no idea where the Aqua classes are coming from

**Reply by MichaelWDenney** [reply]
Hmmm.  Might be some legacy code triggering something...

**Comment by MichaelWDenney** [comment]
Thanks for the suggestions.  I'll take a look.

Our pre-radiance GUI ran on Macs without these issues, so I'm at a bit of loss.  Which is not to say, I think radiance is at fault.  I just don't know where to start.

**Reply by kirill-grouchnikov** [reply]
The biggest difference from core LAFs is probably `RadianceCommonCortex.paintAtScale1x` which gets to the code from Flatlaf to handle high DPI displays at https://github.com/kirill-grouchnikov/radiance/blob/sunshine/common/src/main/java/org/pushingpixels/radiance/common/internal/contrib/flatlaf/HiDPIUtils.java#L39

I wonder if you're seeing similar drawing artifacts in the Flatlaf demo. I know that there were a few similar bugs filed on Flatlaf over the last couple of years in the same area.

**Reply by MichaelWDenney** [reply]
Where can I find the Flatlaf demo?  And do you mean on a Mac or just in general?  I don't have any of the demo code installed on a Mac at the moment.

**Reply by kirill-grouchnikov** [reply]
Their demo is at https://github.com/JFormDesigner/FlatLaf/tree/main/flatlaf-demo

**Comment by MichaelWDenney** [comment]
So, the redraw issues on Macs are due to exceptions being thrown in the middle of repaints because the GUI is trying to use Mac look-and-feel classes that aren't accessible.  For example:

`Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException: Cannot invoke "javax.swing.border.Border.paintBorder(java.awt.Component, java.awt.Graphics, int, int, int, int)" because the return value of "com.apple.laf.AquaMenuPainter$RecyclableBorder.get()" is null`

I assume the Radiance library is not trying to access those classes?

**Reply by MichaelWDenney** [reply]
Also, is there a Radiance look-and-feel that gives more of a Mac friendly look?

**Reply by kirill-grouchnikov** [reply]
There shouldn't be anything in Radiance that is accessing or using (directly or via reflection) anything in the native Aqua look and feel on Macs. There is an old entry in the FAQ (see the very last part of https://github.com/kirill-grouchnikov/radiance/blob/sunshine/docs/faq.md) that talks about how to "mix" Radiance with the native menu bar. I haven't tried that when it was contributed, and I haven't tried that since. 

A couple of things to check in your codebase:

* Check for `UIManager.getSystemLookAndFeelClassName()` anywhere in the startup or later on, and see if there's any manipulation of menu UI delegates before or after Radiance is set
* Check where Radiance is set. Maybe it's before the first frame is created?
* Is there any code path that switches to a different look-and-feel, and maybe misses to update the component tree for some frame?
* Maybe this last point applies to some sort of a cached, offscreen piece of UI that later on is displayed, but is not fully configured to run under Radiance.

As for a Mac friendly look (or Windows friendly look, or Linux friendly look) - no, and no plans. Other active look-and-feels such as FlatLaf and DarkLaf do explore this area, using native API bindings for tighter integration with the title pane, popup menus and some other elements. FlatLaf also has tweaks to configure scroll bars and highlights to kind of get closer to the specific platform. At the present moment, I do not have any plans to integrate any native code into Radiance, or to pursue any platform-specific fidelity.
